[B+C] Added Locale API. Adds BUKKIT-2220, BUKKIT-4507

geNAZt opened the pull request:
##### The Issue:

The Bukkit API has no possibility to access the Settings from the Client. This PR adds a LocaleManager for Plugins and a Event which gets fired when the Client changes its Settings.
##### Justification for this PR:

Adding Language and other Settings to the API will help Plugin Developers providing more customised Texts for the Players.
##### PR Breakdown:

Adding a LocaleManager which holds the loaded Languages for a Plugin
Adding a PlayerSettingsChangeEvent which gets fired when the Client Settings change. It reflects the View Distance, the new Locale and Chat Settings.
Added a ChatMode Enum to reflect all possible states of the Chat
##### Testing Results and Materials:

For testing i made a Little Plugin which loads in two Languages and greets the Player based on his language Setting. Source: https://github.com/geNAZt/LocaleTestPlugin
##### Relevant PR(s):

B-983- https://github.com/Bukkit/Bukkit/pull/983 - The new API for the Locales and Settings
CB-1286 - https://github.com/Bukkit/CraftBukkit/pull/1286 - Extended the EventFactory and the CraftPlayer to get the Locale and Settings
##### JIRA Ticket:

BUKKIT-2220: https://bukkit.atlassian.net/browse/BUKKIT-2220
BUKKIT-4507: https://bukkit.atlassian.net/browse/BUKKIT-4507

Commit:
Added PlayerSettingsChangeEvent to the API to reflect and Settings Change in the Client
Added a getLocale() to the Player Entity Class
Added a Locale Enum to reflect all possible Languages in the Client
Added ChatMode Enum to reflect all possible Settings in the Client
Added a LocaleManager which can be used for translation and Language File Handling
Added getLocaleManager() to the JavaPlugin abstract class to get access to the Plugin specific LocaleManager

Commit:
Fixed the imports here to

Black-Hole commented:
I think the magic values should be CraftBukkit only. Have a look at API added more recently, for instance https://github.com/Bukkit/Bukkit/blob/master/src/main/java/org/bukkit/Sound.java and https://github.com/Bukkit/CraftBukkit/blob/master/src/main/java/org/bukkit/craftbukkit/CraftSound.java

geNAZt commented:
OK thats not part of the contribution Guidelines but i will change it

Commit:
No magic values inside Bukkit API

TheE commented:
I do not really see a need for a LocaleManager. Java already has an extremely powerful standard for handling localizations: [ResourceBundles](http://docs.oracle.com/javase/7/docs/api/java/util/ResourceBundle.html) that automatically matches a loaded resource-bundle to a given locale, pre-caches bundles and allows even further control via custom [Control](http://docs.oracle.com/javase/7/docs/api/java/util/ResourceBundle.Control.html) classes. If anything, bukkit could provide a custom implementation that uses (per-plugin?) yaml-files. This could also be provided on a plugin-level using a library, as I did within [MyWarp](https://github.com/TheE/MyWarp/tree/master/src/main/java/me/taylorkelly/mywarp/localization).

Having this approach in mind, there is also no need for a LocaleChangeEvent or everything similar. Using resource-bundles, you need to get the locale everytime you get a key out of one (as said, caching is handled entirely via the JVM). Plus, if you need to use custom [MessageFormats](http://docs.oracle.com/javase/7/docs/api/java/text/MessageFormat.html) the locale is used again, to get the right format (e.g. currency, date, time...). Both results in getting the locale everytime you send a message.

To, all in all, I do not see a great benefit from choosing a half-backed, less powerful bukkit API over an extremely powerful, well-known Java standard.

geNAZt commented:
Well i could have used ResourceBundles, but what if i want to have another Storage behind them. Also the LocaleManager could be extended to be even wider used (for example use it in combination with ebeans). 

The locale Manager does not support currency (because Minecraft has no money [the economy plugins also set their own prefix/suffix]), date and time can not be used because you don't know the Timezone (for example en_US has 4? Timezones).

Also if you think you approach is better then why dont you share your Code with Bukkit and make your Implementation a part of the API ?

pschichtel commented on src/main/java/org/bukkit/Locale.java:
```diff
@@ -0,0 +1,124 @@
+package org.bukkit;
+
+public enum Locale {
+    //Sorts of english
```
kind of inconsistent: Country prefixed (Australian English) vs Country code as a suffix (English (UK))

Commit:
Made the Enum more consistent

TheE commented:
Take a look at ResourceBundle.Control. It allows complete customization of the actual ResorceBundle itself. You can easily use Yaml as file-type (as I did with MyWarp), you could make use eBeans of JDBC for databases (witch is apparently done by various major enterprise solutions), you can control almost everything.

For the currency part, you got me wrong. What I meant is, that you have a localized string that has some wildcards that are replaced at runtime (the parts in braces are the values, as seen in the MessageFormat java-docs):

```
You still need to wait {0, number, #0.00} minutes before you can use the warp '{1}' again.
```

The first value creates a special DecimalFormat that only shows the two first two digits after the decimal point, the seconds is just a normal string replacement. Given that the string is loaded, all you need to do is setup the MessageFormat:

```
MessageFormat mf = new MessageFormat("You still need to wait {0, number, #0.00} minutes before you can use the warp '{1}' again.");
mf.setLocale(Locale.en_US); // for demonstration, should be dynamic
mf.format(1.20, "yourWarp");
```

Now, the great thing here is, that MessageFormat is actually pretty smart and knows how the needed language expects the formatting. When using the above string with an english locale, you receive:

```
You still need to wait 1.20 minutes before you can use the warp 'yourWarp' again.
```

Using a german locale (with the same sentence!) you will receive:

```
You still need to wait 1,20 minutes before you can use the warp 'yourWarp' again.
```

Notice how it changes the dot to a comma? This also applies to other things such as currency (real currency, not minecraft-stuff), dates or times. The MessageFormat will receive a simple and standard DateFormat and format it the way it is supposed to be displayed in the corresponding locale. And of course, there are CustomFormats that allow complete grammar changes if needed, e.g.:

```
There {0,choice,1#is only one page|1<are only {0, number, integer} pages} in total.
```

All I am saying it that localization is an extremely complicated tasks, that is often taken to lightly. Given that java already provides proper APIs to handle localizations, that these APIs can be customized to meet all needs that can occur and that these APIs are also used in about all enterprise solutions I am aware of, there is no need to come up with our own API, it is just a waist of resources on something, has is already available in a much better state.

geNAZt commented:
Ok then i got you wrong, i will consider to switch to RessourceBundles since they can be configured as wanted.

geNAZt commented:
Well i cant seem to get the ResourceBundles to get running. They keep using the wrong Classloader to read from the  Plugin JAR. If someone wants to look into it i can open a new Branch. But for this PR i will stay with the simple LocaleManager i did here.

ryantheleach commented:
It's a cool idea, but I agree with TheE.

would make it easier for translate plugins to do there job.

geNAZt commented:
Well the ResourceBundles add a bit of things i dont like. For example the basename which i dont need, also its harder to implement because the ResoureBundle.Control Class uses a classloader out of nowhere and it is the wrong. I found no way to bind the PluginClassLoader into the ResourceBundle.Control.

geNAZt commented:
Well i got the ResourceBundle running. Fixing some small Bugs and then i will update this PR one last time.

Commit:
Changed the LocaleManager from using YML Fileloader to ResourceBundles
Now the JVM handles everything was has todo with locales
The Locale Enum was removed because it is no longer needed
The jars can contain lang/lang_<language>_<country>.yml Files for locales

Commit:
Removed the LocaleLoadFailedException. It was only used in the old LocaleManager

pschichtel commented:
Why not use Java's Locale instead of a String?

geNAZt commented:
You mean in the Player interface ? That would lead to a lookup table and a lookup call in CraftBukkit, which i dont want todo since i must have a Manager for the Locales in the API so or so

pschichtel commented:
Yes, in Player and in the event. I'm just talking about the plain Locale object, so you could directly use it for date formatting and similar things.

I'd replace [this](https://github.com/Bukkit/CraftBukkit/pull/1286/files#diff-e117252db4c95f5133ef2658e893ec69R33) with `public Locale locale = Locale.US;`

and [this](https://github.com/Bukkit/CraftBukkit/pull/1286/files#diff-e117252db4c95f5133ef2658e893ec69R930) with `this.locale = LocaleUtils.toLocale(packetplayinsettings.c())`

geNAZt commented:
Nope, dont make changes to systems inside nms when not needed (minimal diff policy) also toLocale is slower as the lookup table.

Test code: https://gist.github.com/geNAZt/7860816
Result:

36517 ms
1037 ms

pschichtel commented:
You would add 1 additional changed line to the diff and I doubt performance is critical here as this code is only ever called when the player changes his settings and you could remove the conversion in other places (haven't checked where you convert it) as well as plugins could directly work with the locale instead of manually converting it (probably) over and over again.

geNAZt commented:
Well the goal for the LocaleManager is never to convert to the Locale. The event should only announce that the Locale has changed. 

But for this change i would need to change two lines in NMS code. I want to wait till a Repo Maintainer/Owner has looked into this PR to check if its ok to change the System of storing the locale inside of NMS. I dont know how far this change would affect the NMS code

TheE commented:
Looks much better now.

Other things that might need some thoughts (or at least troubled me when writing the localization for MyWarp):
1. File encoding. The YamlConfiguration always defaults to the default encoding of the OS it runs on. This might be a problem when dealing special characters (e.g. ü, ä, è, â...) - might need some testing.
2. Maybe it is a better idea to use the plugin's name as a base-name. Or something like "plugin_localization". But thats just my opinion.
3. What about other classes implementing CommandSender (console...)? Unless plugin-developers are expected to handle these conflicts by themselves, it might be a good idea to make the translate-method receive a CommandSender and handle the locale internally (using the default locale for non-players).
4. What about other messaging APIs, namely the whole new Conversation API? Is there anything that needs to be done to make use of the LocaleManager here?

geNAZt commented:
Another thing i found out is that the Client sends the Settings Packet twice (once if i select a new Language in the client and once if i hit "Done" in the Settings page). Should i leave the Vanilla Behaviour ? Should i build a workaround to check if the Settings really have chaged ? Should i report a bug in the Client to Mojang ?

TheE thanks for the additions. I will implement a translate function for ConsoleSenders (which are really consoles) and let them get Messages based on the Default Locale.

The Conversation API could use normal Strings out of the LocaleManager. I think this should not be a problem.

pschichtel commented:
as the events says PlayerSettingsChange it should only be called if the settings actually changed I'd say.

Commit:
Add a translation method for CommandSenders which use the default Locale.
Also added @throws Javadocs for the translate methods because they throw MissingResourceException when the Resource was not found or the key is not present in the Resource

mbax commented:
How would you be able to distinguish between the selection and actual completion (done) of the settings change client side?

Commit:
Made the YamlRessourceBundle UTF-8 compatible

geNAZt commented:
Well i would check what the EntityPlayer from the NMS has stored and if any value differs i would call the Event. Currently it calls the event even if the EntityPlayer has not changed it settings

Commit:
Change to the fancy z (American English)

Commit:
Changed the language Basename (two Plugins will conflict with the old naming)
The files now need to be: lang/<pluginname in lowercase>_<language>_<country>.yml
For example lang/localetestplugin_en_US.yml for my Test Plugin

Commit:
Remove the debug which i needed for finding the conflict Bug

ryantheleach commented:
Is it possible the event is in the wrong location or not reading something if the EntityPlayer from NMS can differ?

Edit: I suspect that the client will resend the packet if you cancel the menu as well, reverting it back to the old settings, this would imply that the settings are applied immediately after clicking the button, so the server should do the same.

it would then double send it if it was the only change made, as other settings that are not immediately applied when changed would need to be sent.

Commit:
Added a reload handle and care of Plugin reload. Now on every Plugin reload the Locale resources of this Plugin also get reloaded.
You can trigger this by using getLocaleManager().reload();

geNAZt commented:
The EntityPlayer from the NMS cant differ from the Event State. If you catch all Events you can be 100% sure that you got the current State of the EntityPlayer from NMS.

Edit:
Yes the problem is if i open up the "Language" Selector in the Settings and i click on another Language the Client sends of a Packet to the server. I think it would be right to send a Packet if the User clicks on "Done", "Cancel" or closes the Client.

ryantheleach commented:
I figured out what was happening afterwards hence the edit.

geNAZt commented:
I will remove the ResoureBundle thing because it sucks. I will try to stay with the MessageFormat things because they are nice. I will implement my own Loading of Files.

Commit:
New ResourceLoader interfaces
New ResourceManager
Made LocaleManager compatible with new ResourceManager

Now its up to the Plugin Devs which file should be loaded with which Loader into which Locale. The Plugin Dev can register its own ResourceLoader implementations to load from even more Formats.

The loading of Language Files is now: getLocaleManager().load(Locale.US, "lang/en_US.yaml");. This example would be loaded with the builtin YML Resource Loader.

Commit:
Improve Javadocs

Commit:
Added PropertiesResourceLoader so you also can load .properties as Language Files

Commit:
Added and fixed some tests

Commit:
Added cleanup() and reload() Tests
Fixed a bug that corrects the reloading

Commit:
Added a test to test the changing of Locales while the LocaleManager is running

geNAZt commented:
Added a Fix for the multiple PlayerSettingsChangeEvent in the CB PR. Please also look there :)

TheE commented:
I clearly hope that this is not merged in the current form. Not using solid, flexible and well-known standards because they "suck" is an absolutely terrible idea, especially if the replacements is nowhere as flexible and robust as the standard itself:
- The LocalManager here does not provide a proper child/parent relationship at it is not able to automatically match nested localizations from different files as Java's RessourceBundle does.  
  It is just a waist of resource to hold a complete bundle for en_GB (Great Britain) and an exact copy for en_US (USA). And you do not just need to implement subformats like en (english), you also need to make the key-lookup work recursively: If you want to modify just a few values for en_US (lets say, replace 'colour' by 'color'), there is absolutely no point in having a complete localization-file for en_US, when, except for the few modified values, everything is exactly the same as in the parental file, en.
- The ResourceLoader cache implemented in the LocalManager here is not able to provide a proper live-cycle-management as ResourceBundle.Control could. It is not even able to automatically remove cache ResourceLoaders when they block memory that is needed otherwise, which Java's ResourceBundle.Control  does automatically.
- The stored ResourceLoaders here are not thread-safe, Java ResourceBundles are, once they are constructed.

As I said before, localizations are a complicated task that cannot be taken to lightly. I do not see a single point that stands against using Java's standard (except, maybe that it needs more work to write the implementation for bukkit, but naturally that is no valid argument), but multiple points that stand against using the custom API you provided.

geNAZt commented:
Well RessourceBundles have some big cons which are a no go in Systems like Bukkit. RessourceBundles have no seperation model so you must use one RessourceBundle.Control for the whole cache. What means that you cant load in a ressource twice which would lead to a System which relies on dynamic basenames. And to support this all Locale Files need to be prefixed with the pluginname or something like that.

Also proper reloading of parts of the RessourceBundle cache is nearly impossible. You have to check the Basename and the time at which the Ressource entered the cache to terminate if it needs to be reloaded or not. Also the RessourceBundle implementation does not support unloading of certain Ressources (like needed if plugins get unloaded). 

So this leads me to the conclusion that the RessourceBundle itself maybe a good System if you build Java Application which need a Localisation but for a dynmaic Plugin driven API its not good because of its global State of the Cache.

So i came up with the Idea to have Ressources splitted into their Plugins. What you see in this PR is a Implementation of Ressourceloading which allows all the things stated above. I agree that some features are leaft out as you stated, but it can be completed for that points. 

For thread safety there is to say that you can get the keys in every Thread you want. Only the loading / reloading of Locale Files need to be done in the Mainthread. I dont see a problem in this.

Commit:
Added the ability to inherit Languages. So you can load for example a "en" Locale and a "en_US" locale. If the key is not present in "en_US" the "en" locale gets used.

Commit:
All ResourceLoaders are SoftReferences now so the GC can clear them out if the JVM needs more memory.

Commit:
Added more Thread safety

Commit:
Added a shared lock to each ResourceManager so you can be sure that only one Thread can write into it

TheE commented:
As far as I understand the java-docs, you can perfectly manage resource-bundles per class-loader. The static method `ResourceBundle.clearCache(ClassLoader loader)` only clears the cache for bundles loaded with the given class-loader (as does `ResourceBundle.clearCache()` using the class-loader that calls the method). Using the plugin's class-loader here should have the desired results.

I also do not see anything that blocks using different ResourceBundle.Control implementations. The static method `ResourceBundle.getBundle(String baseName, Locale targetLocale, ResourceBundle.Control control)` that returns a loaded resource-bundle to work with can be called using different ResourceBundle.Control implementations. A proper bukkit-implementation should provide a default ResourceBundle.Control that uses yaml-files in the plugin directory, but would also allow plugins to provide their own implementation of ResourceBundle.Control for custom files (sql-databses, xml-files, plain class-files etc.)

geNAZt commented:
"ResourceBundle.clearCache(ClassLoader loader)" clearsthe complete cache from ALL loaded Resources. Even the ones that arent owned by the Plugin. As you see in the Commit History i tried to implement the ResourceBundles via seperate Classloaders. So you have one ResourceBundle.Control for every Classloader (and therefore you can control the Cache via Classloaders right). "Using the plugin's class-loader here should have the desired results.", this is not so as you can see in the Commit History. If two different Plugins load in the same Resource via different Classloaders they overwrite each other.

Well you can ofc use different ResourceBundle.Control. But the problem is that the ResourceBundle Cache is global. So you can again only load one basename once. And i find Resource Names like "thisismyplugin_en_US.yml" very ugly.

TheE commented:
All I see in the commit history is `resourceBundleControl.reloadResources()` which not the same approach as using `ResourceBundle.clearCache(ClassLoader loader)`. While you could have one instance of ResourceBundle.Control per plugin (or even more, if needed), it does not help with removing cached bundles that are all in the same cache.

geNAZt commented:
"ResourceBundle.clearCache(ClassLoader loader)" you can go and try it. I did it and it removed ALL Resources loaded under the Plugin Classloader (i think somehow the Plugins get loaded by the same loader).

geNAZt commented:
Any status of this PR ?

jessestricker commented:
I really do like #899 much more, as it is not that much of an overload. It is easy-to-use and simple. Developers could implement their own translations manager, using e.g. ResourceBundles or something self-made, like what @geNAZt did.
IMO a translation API in this size not needed, so I would appreciate it if it won't get pulled.

geNAZt commented:
Well its better to have a global "standart" for things like this. It is plugable with your own extensions and can be used in an easy way. I dont see why all Developers should be forced to create a Implementation of a Language System when the API could provide one.

You also can make your own implentation of Language since this PR throws out Events on Locale changes and opens the getLocale() to the API. So i dont see a valid point in your Comment to not merge this.

jessestricker commented:
So are you going to write a documentation on this? Because new devs are not going to know how this works without  looking into your code. They may know Java's ResourceBundles and use that instead.

geNAZt commented:
Yes the Language API in here is documented like any other API in Bukkit. For example:

``` java
 /**
  * Translate the Text based on the Player locale.
  * If the locale from the player is not loaded the LocaleManager will try to load it, if this fails
  * it will use the default Locale. If this is also not loaded you will get a ResourceNotLoadedException
  *
  * @param player Player from which the Locale should be used
  * @param translationKey The key in the ResourceLoader which should be translated
  * @param args The Arguments which will be passed into the String when translating
  * @return The translated String
  * @throws ResourceNotLoadedException when the Resource for the locale could not be loaded or the key is missing
  * @throws ResourceLoadFailedException when the GC has cleared out the ResourceLoader and it could not be reloaded into the cache
  */
```

For translating a Message out of the Resource for the given Locale and passing in the arguments

natemort commented on src/main/java/org/bukkit/ChatMode.java:
```diff
@@ -0,0 +1,7 @@
+package org.bukkit;
+
+public enum ChatMode {
```
You should have documentation describing what this enum does along with each value.

natemort commented on src/main/java/org/bukkit/entity/Player.java:
```diff
@@ -721,4 +721,11 @@
+     *
+     * @return the Locale which has been set
+     */
+    public String getLocale();
```
What's the standard for the locale strings?  What should I expect one to look like?  (This should be documented).

Commit:
Improved Docs for getLocale() and the ChatMode Enum

Commit:
Improved Docs for getLocale() and the ChatMode Enum

geNAZt commented:
Is there any progress in here ? I fixed the Docs 6 days ago and nothing seems to happen here.

turt2live commented:
At a quick glance I see a merge conflict. Haven't looked farther than that though

Commit:
Merge pull request #1 from Bukkit/master

Merge current Bukkit Master

Commit:
Merge branch 'master' into LocaleAPI

Conflicts:
	src/main/java/org/bukkit/plugin/java/JavaPlugin.java

geNAZt commented:
Should be ok now

geNAZt commented:
Is there any Progress here ? I fix things and then nobody cares of it...... Is this getting reviewed or not ?

storm345 commented:
This is such a great idea I'm surprised it wasn't pulled instantly!

pschichtel commented:
What I've seen: settings like the chat mode and the view distance are not accessable in the Player interface, but only though the event. Maybe introduce something like a PlayerSettings object that makes these settings accessible.

Wolvereness commented:
@storm345 https://github.com/Bukkit/Bukkit/pull/983#issuecomment-31551668

> This is such a great idea I'm surprised it wasn't pulled instantly!

You really shouldn't be surprised that someone's first draft is not quite ready to be pushed live to the plethora of servers running CraftBukkit.

geNAZt commented:
@Wolvereness https://github.com/Bukkit/Bukkit/pull/983#issuecomment-31552501

So what is wrong with this that it can't be pushed to the plethora of servers ?

Wolvereness commented:
@geNAZt https://github.com/Bukkit/Bukkit/pull/983#issuecomment-31558075

> @Wolvereness https://github.com/Bukkit/Bukkit/pull/983#issuecomment-31552501
> 
> So what is wrong with this that it can't be pushed to the plethora of servers ?

Bare minimum, there are missing Javadocs. Additionally, you should be using proxy classes for interface tests (as demonstrated by some of the CraftBukkit tests). Touching those files every time a method is added becomes a (avoidable) pain.

There is additional review required that I can't personally do at this exact moment in time, other than responding to some very obvious points.

Commit:
Improved JavaDocs

geNAZt commented:
I improved the JavaDocs. But i never worked with Java Proxies. I tried to do this but i failed. Sorry

Wolvereness commented:
Basically, copy+paste https://github.com/Bukkit/CraftBukkit/blob/master/src/test/java/org/bukkit/support/DummyServer.java but change the reflected methods.

Commit:
Changed testing to use Proxies to avoid changes if the Interfaces change

geNAZt commented:
Ok made it. It works. Thanks @Wolvereness

geNAZt commented:
And again 5 days without any progress......

Wolvereness commented:
@geNAZt https://github.com/Bukkit/Bukkit/pull/983#issuecomment-31855736

> And again 5 days without any progress......

https://twitter.com/EvilSeph/status/420612986198585344

geNAZt commented:
Yeah thanks for this Information. Maybe there should be a central Place for PR/Code freezes/Policy stuff... I spoke with EvilSeph about it and he will start something in his Website. But maybe you should implement something in your Project which Informs about the current Status ? Like a Devtracker or a RSS, Mailing list or something like that.

But the support for PRs is a bit meh. Sorry if this sounds a bit hard but thats only my point of view.

geNAZt commented:
Is there any progress in reviewing this ?

turt2live commented:
@geNAZt if you, or anyone else, is still interested in this please address the following concerns:
- The ability to be pulled (update to master)

Thanks!

pschichtel commented:
I could update this if @geNAZt is not going to do it.
